samplers/aws/xray/internal/client.go

```diff
@@ -0,0 +1,200 @@
+}
+
+// samplingStatisticsDocument is used to store current state of sampling data.
+type samplingStatisticsDocument struct {
+	// a unique identifier for the service in hexadecimal
+	ClientID *string
+
+	// the name of the sampling rule
+	RuleName *string
+
+	// the number of requests that matched the rule
+	RequestCount *int64
+
+	// the number of requests borrowed
+	BorrowCount *int64
+
+	// the number of requests sampled using the rule
+	SampledCount *int64
+
+	// the current time
+	Timestamp *int64
+}
+
+// getSamplingTargetsOutput is used to store parsed json sampling targets
+type getSamplingTargetsOutput struct {
+	LastRuleModification    *float64                  `json:"LastRuleModification,omitempty"`
+	SamplingTargetDocuments []*samplingTargetDocument `json:"SamplingTargetDocuments,omitempty"`
+	UnprocessedStatistics   []*unprocessedStatistic   `json:"UnprocessedStatistics,omitempty"`
+}
+
+// samplingTargetDocument contains updated targeted information retrieved from X-Ray service
+type samplingTargetDocument struct {
+	// the percentage of matching requests to instrument, after the reservoir is
+	// exhausted
+	FixedRate *float64 `json:"FixedRate,omitempty"`
+
+	// the number of seconds for the service to wait before getting sampling targets
+	// again
+	Interval *int64 `json:"Interval,omitempty"`
+
+	// the number of requests per second that X-Ray allocated this service
+	ReservoirQuota *int64 `json:"ReservoirQuota,omitempty"`
+
+	// when the reservoir quota expires
+	ReservoirQuotaTTL *float64 `json:"ReservoirQuotaTTL,omitempty"`
+
+	// the name of the sampling rule
+	RuleName *string `json:"RuleName,omitempty"`
+}
+
+type unprocessedStatistic struct {
+	ErrorCode *string `json:"ErrorCode,omitempty"`
+	Message   *string `json:"Message,omitempty"`
+	RuleName  *string `json:"RuleName,omitempty"`
+}
+
+type xrayClient struct {
+	// http client for sending sampling requests to the collector
+	httpClient *http.Client
+
+	// resolved URL to call getSamplingRules API
+	samplingRulesURL string
+
+	// resolved URL to call getSamplingTargets API
+	samplingTargetsURL string
+}
+
+// newClient returns an HTTP client with proxy endpoint.
+func newClient(addr string) (client *xrayClient, err error) {
+	endpoint := "http://" + addr
+
+	endpointURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// construct resolved URL for getSamplingRules and getSamplingTargets API calls
+	samplingRulesURL := endpointURL.String() + "/GetSamplingRules"
+	samplingTargetsURL := endpointURL.String() + "/SamplingTargets"
+
+	return &xrayClient{
+		httpClient:         &http.Client{},
+		samplingRulesURL:   samplingRulesURL,
+		samplingTargetsURL: samplingTargetsURL,
+	}, nil
+}
+
+// getSamplingRules calls the collector(aws proxy enabled) for sampling rules.
+func (c *xrayClient) getSamplingRules(ctx context.Context) (*getSamplingRulesOutput, error) {
+	samplingRulesInput, err := json.Marshal(getSamplingRulesInput{})
```

Comment:
Yeah `nextToken` field is never used and the API expects nil value of that but I have created this struct to use more like a place holder struct for this. Maybe in future if API changes and if we want to leverage this value inside this implementation then we can without introducing anything at that time. I don't see any overhead on keeping it in a struct like this. I think the API without input params looks more confusing so maybe we can just keep it like this but happy to hear your thoughts if you think otherwise.